psi4/src/psi4/libmints/mintshelper.cc

```diff
@@ -894,34 +905,96 @@ SharedMatrix MintsHelper::ao_eri_shell(int M, int N, int P, int Q) {
+    }
+    return ao_helper("AO F12 Tensor", ints_vec);
 }
 
 SharedMatrix MintsHelper::ao_f12_squared(std::vector<std::pair<double, double>> exp_coeff) {
-    std::shared_ptr<TwoBodyAOInt> ints(integral_->f12_squared(exp_coeff));
-    return ao_helper("AO F12 Squared Tensor", ints);
+    std::vector<std::shared_ptr<TwoBodyAOInt>> ints_vec(nthread_);
+    ints_vec[0] = std::shared_ptr<TwoBodyAOInt>(integral_->f12_squared(exp_coeff));
+    for (size_t i = 1; i < nthread_; i++) {
+        ints_vec[i] = std::shared_ptr<TwoBodyAOInt>(ints_vec.front()->clone());
+    }
+    return ao_helper("AO F12 Squared Tensor", ints_vec);
 }
 
 SharedMatrix MintsHelper::ao_f12_squared(std::vector<std::pair<double, double>> exp_coeff,
                                          std::shared_ptr<BasisSet> bs1, std::shared_ptr<BasisSet> bs2,
                                          std::shared_ptr<BasisSet> bs3, std::shared_ptr<BasisSet> bs4) {
     IntegralFactory intf(bs1, bs2, bs3, bs4);
-    std::shared_ptr<TwoBodyAOInt> ints(intf.f12_squared(exp_coeff));
-    return ao_helper("AO F12 Squared Tensor", ints);
+    std::vector<std::shared_ptr<TwoBodyAOInt>> ints_vec(nthread_);
+    ints_vec[0] = std::shared_ptr<TwoBodyAOInt>(intf.f12_squared(exp_coeff));
+    for (size_t i = 1; i < nthread_; i++) {
+        ints_vec[i] = std::shared_ptr<TwoBodyAOInt>(ints_vec.front()->clone());
+    }
+    return ao_helper("AO F12 Squared Tensor", ints_vec);
+}
+
+SharedMatrix MintsHelper::ao_f12g12(std::vector<std::pair<double, double>> exp_coeff) {
+    std::vector<std::shared_ptr<TwoBodyAOInt>> ints_vec(nthread_);
+    ints_vec[0] = std::shared_ptr<TwoBodyAOInt>(integral_->f12g12(exp_coeff));
+    for (size_t i = 1; i < nthread_; i++) {
+        ints_vec[i] = std::shared_ptr<TwoBodyAOInt>(ints_vec.front()->clone());
+    }
+    return ao_helper("AO F12G12 Tensor", ints_vec);
+}
+
+SharedMatrix MintsHelper::ao_f12g12(std::vector<std::pair<double, double>> exp_coeff,
+                                    std::shared_ptr<BasisSet> bs1, std::shared_ptr<BasisSet> bs2,
+                                    std::shared_ptr<BasisSet> bs3, std::shared_ptr<BasisSet> bs4) {
+    IntegralFactory intf(bs1, bs2, bs3, bs4);
+    std::vector<std::shared_ptr<TwoBodyAOInt>> ints_vec(nthread_);
+    ints_vec[0] = std::shared_ptr<TwoBodyAOInt>(intf.f12g12(exp_coeff));
+    for (size_t i = 1; i < nthread_; i++) {
+        ints_vec[i] = std::shared_ptr<TwoBodyAOInt>(ints_vec.front()->clone());
+    }
+    return ao_helper("AO F12G12 Tensor", ints_vec);
+}
+
+SharedMatrix MintsHelper::ao_f12_double_commutator(std::vector<std::pair<double, double>> exp_coeff) {
+    std::vector<std::shared_ptr<TwoBodyAOInt>> ints_vec(nthread_);
+    ints_vec[0] = std::shared_ptr<TwoBodyAOInt>(integral_->f12_double_commutator(exp_coeff));
+    for (size_t i = 1; i < nthread_; i++) {
+        ints_vec[i] = std::shared_ptr<TwoBodyAOInt>(ints_vec.front()->clone());
+    }
```

Comment:
This might actually be better to run in parallel. That way the memory can get allocated to the processor's local memory.